src/blob/index/bptree/core.rs

```diff
@@ -0,0 +1,283 @@
+/// 6. Non-leaf nodes of btree in format
+///    `NodeMeta | keys_arr | pointers_arr`,
+///    where pointer is offset in file of underlying node with searched key
+use super::prelude::*;
+
+pub(super) const BLOCK_SIZE: usize = 4096;
+
+#[derive(Debug, Clone)]
+pub(crate) struct BPTreeFileIndex {
+    file: File,
+    header: IndexHeader,
+    metadata: TreeMeta,
+}
+
+#[async_trait::async_trait]
+impl FileIndexTrait for BPTreeFileIndex {
+    async fn from_file(name: FileName, ioring: Option<Rio>) -> Result<Self> {
+        trace!("open index file");
+        let file = File::open(name.to_path(), ioring)
+            .await
+            .context(format!("failed to open index file: {}", name))?;
+        let header = Self::read_index_header(&file).await?;
+        let metadata = Self::read_tree_meta(&file, &header).await?;
+
+        Ok(Self {
+            file,
+            header,
+            metadata,
+        })
+    }
+
+    async fn from_records(
+        path: &Path,
+        ioring: Option<Rio>,
+        headers: &InMemoryIndex,
+        filter: &Bloom,
+        recreate_index_file: bool,
+    ) -> Result<Self> {
+        let res = Self::serialize(headers, filter)?;
+        let (mut header, metadata, buf) = res;
+        clean_file(path, recreate_index_file)?;
+        let file = File::create(path, ioring)
+            .await
+            .with_context(|| format!("file open failed {:?}", path))?;
+        file.write_append(&buf).await?;
+        header.written = 1;
+        let serialized_header = serialize(&header)?;
+        file.write_at(0, &serialized_header).await?;
+        file.fsyncdata().await?;
+        Ok(Self {
+            file,
+            metadata,
+            header,
+        })
+    }
+
+    fn file_size(&self) -> u64 {
+        self.file.size()
+    }
+
+    fn records_count(&self) -> usize {
+        self.header.records_count
+    }
+
+    async fn read_filter(&self) -> Result<Bloom> {
+        trace!("load filter");
+        let mut buf = vec![0; self.header.filter_buf_size];
+        trace!("read filter into buf: [0; {}]", buf.len());
+        self.file
+            .read_at(&mut buf, self.header.serialized_size()? as u64)
+            .await?;
+        Bloom::from_raw(&buf)
+    }
+
+    async fn find_by_key(&self, key: &[u8]) -> Result<Option<Vec<RecordHeader>>> {
+        let root_offset = self.metadata.root_offset;
+        let mut buf = [0u8; BLOCK_SIZE];
+        let leaf_offset = self.find_leaf_node(key, root_offset, &mut buf).await?;
+        if let Some((fh_offset, amount)) =
+            self.find_first_header(leaf_offset, key, &mut buf).await?
+        {
+            let headers = self.read_headers(fh_offset, amount as usize).await?;
+            Ok(Some(headers))
+        } else {
+            Ok(None)
+        }
+    }
+
+    async fn get_records_headers(&self) -> Result<(InMemoryIndex, usize)> {
+        let mut buf = self.file.read_all().await?;
```

Comment:
On start it will create lots of iops to read whole indexes and will result in slow start. Should we consider making it optional?

Reply:
I don't understand you. How we can make it optional?
Current version of pearl calls this method only once during init: for active blob if there is one. We can make active blob optional for `init_from_existing`, in this case pearl will be read-only, but there won't be any `get_records_headers` operations

Reply:
In older version of indices they consisted of header, filter, and record headers. New version (bptree) consists of header, meta (I changed bloom filter of Vec<u8> and make serialization of filter on higher level, because file index doesn't need to know about it), record headers and b+ tree. New version size ~ 1.2 * size of simple indices version

In older version we read the whole file for hash validation. Here I do the same thing (check validate_header function: it also validates if the hash in header is correct). Do you want to make hash check optional OR to have 2 different hashes (for record headers array and for tree)?

Reply:
It would be nice to check how it's done by other storage systems. Simple solution - make it optional. But would it be nice to check how it's done by others.

Reply:
Some FS have CRC on block level (e.g. ZFS, btrfs), but some do not (e.g. ext4)

Reply:
Postgresql added block level checksum since 9.3 https://postgreshelp.com/postgresql-checksum/

Reply:
@agend So, as I understood from your link, postgres stores per page checksum and during loading pages to shared buffer from disk it validates every loaded page. Postgres does that per page, because pages can be loaded independently. If modifications are done with any page (as I understand, they can be done only in shared buffer), checksum will be recalculated for this page when it will be removed from shared buffer (to OS cache or immediately to disk, it depends on behavior of OS and `fsync` usage in postgres implementation). Let's name postgres page as _exchange unit_ (portion of data, which is transfered between disk and shared buffer, and which can be modified (modification triggers checksum recalculation))

In our case, Pearl allows to load **all** key-record_header pairs in memory (`get_record_headers`; and then data can be modified, aka `shared buffer`) and write only **all** (at the same time I mean) key-record_header pairs (`from_records`; data can't be modified, aka `disk` but with read capability). So _exhange unit_  between InMemoryIndex (aka _Pearl shared buffer_) and disk is the whole index tree for blob. 

Did I misunderstand something? If I am not, I think, it's better leave checksum for the whole index tree, because we don't manipulate data (actually, we do, when we retrive elements array by key, but we don't modify them and we may validate checksums on startup for all disks (seems like it's not implemented yet, we only control hash during load in memory, so it is done only for active blob) and have guarantees, that it'll remain correct for all time) so if we'll add checksums for pages, we'll won't get any profit but instead we'll hit performance.

Reply:
hm, there is unnecessary read of bptree, when we load record_headers (~ 25%), and actually, we can add second checksum for bptree (we validate it only at startup, and we don't need to validate it during retrieving of record headers). Is it what you meant?

Reply:
sorry, I missed validation on startup:
While moving indices to disk, Pearl do 2 steps:
1. for serialized data it counts hash, writes this hash into header, and then writes header with serialized data into file
2. sets `written` byte in header, and writes new header in place of the old one

During startup `written` byte is checked (file still may be corrupted in case, when dirty page with written byte (with header, which also include right hash) was flushed to disk, but other pages - weren't; in this case written validation would be true, but hash validation would be false)

Reply:
If there are `N` child nodes at each BPlusTree level, then the calculation will be: `1 + 1/N + 1/(N^2) + ...`. With `N = 16` result will be ~1.07. So there is no problem

Reply:
@ikopylov On practice there are some nuances (alignment of root node, meta of every node, tree meta also need some place) BUT I remembered that my calculations are related to version, when leaves weren't headers themselves, but 'pointers' (offset in file) on headers, so at current version amount of additional size is reduced much (and I guess it will be no more than 7%, as you calculated)
If you are interested in exact value please write about this, otherwise resolve this conversation